types/rdfjs__data-model/index.d.ts

```diff
@@ -0,0 +1,22 @@
+// Type definitions for @rdfjs/data-model 2.0
+// Project: https://github.com/rdfjs-base/data-model
+// Definitions by: Jesse Wright <https://github.com/jeswr>
+//                 Ruben Taelman <https://github.com/rubensworks>
+//                 Andy Davison  <https://github.com/andydavison>
+// Definitions: https://github.com/DefinitelyTyped/DefinitelyTyped
+// Minimum TypeScript Version: 3.9
+
+import * as RDF from "@rdfjs/types";
+
+interface Factory<OutQuad extends RDF.BaseQuad = RDF.Quad, InQuad extends RDF.BaseQuad = OutQuad>
+  // Required is needed here because variable is optional in the spec
+  extends Required<RDF.DataFactory<OutQuad, InQuad>> {
+  // Enable same behavior of version 1 where quads default to RDF.Quad (where possible)
+  quad<Q extends OutQuad = RDF.Quad extends OutQuad ? RDF.Quad : OutQuad>(
+      subject: Q['subject'], predicate: Q['predicate'], object: Q['object'], graph?: Q['graph']): Q;
```

Comment:
Similarly here, I do not think the method should be generic since it always returns quads but should arguably accept any `BaseQuad`. 

```suggestion
  quad(subject: RDF.BaseQuad['subject'], predicate: RDF.BaseQuad['predicate'], object: RDF.BaseQuad['object'], graph?: RDF.BaseQuad['graph']): RDF.Quad;
```

Might also compare with [rdf-ext](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/types/rdf-ext/lib/DataFactory.d.ts#L21) and use `RDF.Quad_Subject` etc.

Reply:
I don't follow on this one, if I have

```ts
factory.quad(
  factory.namedNode('http://example.org/subject'),
  factory.blankNode('34'),
  factory.namedNode('http://example.org/object'),
);
```

Then with this suggestion, there would be no type errors, and it would return an object of type `RDF.Quad`; despite the predicate not being a `Quad_Predicate` since it is a blank node. This is why I made it a generic so that by default, there is a type error in trying to do this

```ts
factory.quad(
  factory.namedNode('http://example.org/subject'),
  factory.blankNode('34'), // Error on this input
  factory.namedNode('http://example.org/object'),
);
```

But one could also just have it emit this as a valid base quad

```ts
factory.quad<RDF.BaseQuad>(
  factory.namedNode('http://example.org/subject'),
  factory.blankNode('34'), // Error on this input
  factory.namedNode('http://example.org/object'),
);
```

Reply:
Gotcha. I'm not sure how practical that is but I see what you want to achieve. Ok, I rest my case. As long as we have the defaults it should be transparent.

Reply:
Resolving per https://github.com/DefinitelyTyped/DefinitelyTyped/pull/57899#discussion_r776648895